File: Array/Greatest Number/By iterative.java
// Method - Iterative(Optimised)
// Time Complexity T - O(n)
// Space Complexity S - O(n) 

class Solution {
    public List<Boolean> kidsWithCandies(int[] candies, int extraCandies) {
        
        // variable to store the maxmium value present in the array candies
        int maxVal = 0;
        
        // loops run for all the values of candies
        for(int i : candies){
            
            // if i is greater than maxVal
            // then assign the "i" to maxVal
            if( i > maxVal){
                maxVal = i; 
            }
        }
        
        // list to store the boolean values 
        List<Boolean> result = new ArrayList<Boolean>();
        
        // loop runs for 0 to n-1
        for(int i = 0; i < candies.length ; i++){
            
            // if the sum of candies[i] and extraCandie is equal to or greater then maximum value of Candies
            // then add boolean value "true" for the respective candies
            if(candies[i]+extraCandies >= maxVal){
                result.add(true);
            }
            // if then sum is less than max
            // then add boolean value "false" for the respective candies
            else{
                result.add(false);
            }
        }
        
        return result;
    }
}

File: Array/First Missing Positive Number/By iterative.cpp
// Method - Iterative(Optimised)
// Time Complexity T - O(n)
// Space Complexity S - O(1)

class Solution {
public:
    int firstMissingPositive(vector<int>& nums) {
       
        // ensure the nums is not empty
        if (nums.size() == 0){
            return 1;
        }    
        
        // loop runs for 0 to n-1 times
        for(int i = 0; i < nums.size();){
            
            // check nums[i] is not greater than size of num
            // or nums[i] values is not negative
            // if so increment "i"
            if(nums[i] > nums.size() || nums[i] <= 0 ){
                i++;
            }
            
            // check nums[nums[i]-1] is equal to nums[i]
            else if(nums[nums[i]-1] == nums[i]){
                i++;
            }
            // else swap the numbers in correct order
            else{
                int temp = nums[nums[i]-1];
                nums[nums[i]-1] = nums[i];
                nums[i] = temp;
            }
        }
        
        int i=0;
        
        // loop rus for o - "n"
        for(i=0;i<nums.size(); i++){
            // check the nums[i]= i+1
            // if not then return i+1
            if (nums[i] != i+1){
                return i+1;
            }
        }
        return nums[i-1]+1;  
    }
};
